apps/system/flash_test/flash_test.c

```diff
@@ -59,80 +59,553 @@
+	buffer = (FAR uint32_t *)malloc(BUF_SIZE);
+
+	if (!buffer) {
+		printf("ERROR: failed to allocate a sector buffer\n");
+		fflush(stdout);
+		goto out_error;
+	}
+	rd_buffer = (FAR uint32_t *)malloc(BUF_SIZE);
+	if (!buffer) {
+		printf("ERROR: failed to allocate a sector buffer\n");
+		fflush(stdout);
+		free(buffer);
+		goto out_error;
+	}
+	/* Open the master MTD FLASH character driver for writing */
 
+	fd = open(blockname, O_RDONLY);
+	if (fd < 0) {
+		printf("ERROR: open /dev/mtd%d failed: %d\n", g_partno, errno);
+		fflush(stdout);
+		goto out_free_buffer;
+	}
+
+	/* Now Read the offset into every block */
+	printf("Reading from media: buff Size %d, times %d\n", BUF_SIZE, (SIZE_1MB/BUF_SIZE));
+
+	offset = 0;
+	systick_time_ms = 0;
+	/* Fill the block with the offset */
+
+	fill_buffer(buffer);
+
+	/* And write it using the character driver */
+	printf("lldb: BCH Read start\n");
+	for (int j = 0; j < (SIZE_1MB/BUF_SIZE); j++) {
+		sys_start_time = get_time();
+		nbytes = read(fd, rd_buffer, (size_t)(BUF_SIZE));
+		if (nbytes < 0) {
+			printf("ERROR: Read to /dev/mtd%d failed: %d\n", g_partno, errno);
+			fflush(stdout);
+			close(fd);
+			goto out_free_buffer;
+		}
+		sys_end_time = get_time();
+		systick_time_ms += sys_end_time - sys_start_time;
+		if (memcmp(rd_buffer, buffer, nbytes)) {
+			printf("ERROR: Read value comparison failed to /dev/mtd%d failed: %d\n", g_partno, errno);
+			fflush(stdout);
+			close(fd);
+			goto out_free_buffer;
+		}
+	}
+	printf("lldb: BCH Read End\n");
+	close(fd);
+	printf("\nSystick time taken for Flash bch Read %ld(ms)\n", systick_time_ms);
+	goto out;
+
+out_free_buffer:
+	free(buffer);
+	free(rd_buffer);
+out_error:
+	return ERROR;
+out:
+	free(buffer);
+	free(rd_buffer);
+	return OK;
+}
+
+static int bch_erase(flash_t flash)
+{
+	int ret;
+	int fd;
+	int g_partno = 0;
+
+	if (flash == FLASH_INTERNAL) {
+		g_partno = INT_FLASH_PART;
+	} else if (flash == FLASH_EXT) {
+		g_partno = EXT_FLASH_PART;
+	}
+	fd = open(blockname, O_RDWR);
+	if (fd < 0) {
+		printf("ERROR: open failed for: %s\n", blockname);
+		fflush(stdout);
+		return ERROR;
+	}
+	sys_start_time = get_time();
+	printf("lldb: BCH Erase start\n");
+	ret = ioctl(fd, MTDIOC_BULKERASE, 0);
+	if (ret < 0) {
+		fdbg("ERROR: ioctl ERASE failed: %d\n", ret);
+		close(fd);
+		return ERROR;
+	}
+	printf("lldb: BCH Erase End\n");
+	sys_end_time = get_time();
+	printf("\nSystick time taken for Flash bch Erase %ld(ms)\n", (sys_end_time - sys_start_time));
+	close(fd);
+	return OK;
+}
+
+#ifdef CONFIG_BUILD_FLAT
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
-static int flash_read(unsigned long addr, int size)
+static int mtd_read(unsigned long addr, int size, flash_t flash)
 {
 	int ret;
 	int offset;
 	int buf_idx;
 	int read_size;
 	int remaining;
-	FAR uint8_t *buffer;
+	FAR uint32_t *buffer;
+	FAR uint8_t *rd_buffer;
```

Comment:
I understand what `rd_buffer` is from the name. But I don't know what `buffer` is. Could you rename it?

Reply:
rename the buffer -> wr_buffer.